### PR TITLE
Add strict option to `MapAxis.downsample`

### LINKED ISSUE #5047: Modify `MapAxis.downsample`
Modify `MapAxis.downsample` to have an option strict=False option which when set to True will simply put the remainder in the last bin
 https://github.com/gammapy/gammapy/pull/5007#discussion_r1455938626_

### PR DESCRIPTION
**Pull request by Astro-Kirsty**
This PR resolves #5047

Adding a strict option to the downsample, such that when it is true, it places the remainder in the last bin

### CODE REVIEW

**Review by QRemy — COMMENTED** [review]
Thanks, @Astro-Kirsty, intuitively I would have expected that `strict=True` impose that the number bins is strictly divisible by factor.
Do you plan to have this used also in the Map.dowmsample ? because having the domsampling factor different for the last bins will break the code there.

File: gammapy/maps/axes.py [hunk 1]
```diff
@@ -1122,6 +1122,10 @@ def downsample(self, factor):
         ----------
         factor : int
             Downsampling factor.
+        strict : bool
+
+            If True, then the remainder is placed in the last bin
```
Comment (AtreyeeS) [inline gammapy/maps/axes.py:1127]: The current docstring `requires nbin to be dividable by the factor..`, must be adapted.
I guess the behaviour should be
- `strict=True` : nbin must to be divisible by the factor 
- `strict=False` the reminder bins are put in the last bin of the new axis

File: gammapy/maps/axes.py [hunk 2]
```diff
@@ -1131,23 +1135,38 @@ def downsample(self, factor):
         if self.node_type == "edges":
             nbin = self.nbin / factor
 
-            if np.mod(nbin, 1) > 0:
-                raise ValueError(
-                    f"Number of {self.name} bins is not divisible by {factor}"
-                )
+            if strict and np.mod(nbin, 1) > 0:
```
Comment (AtreyeeS) [inline gammapy/maps/axes.py:1138]: this should be `strict=False`

File: gammapy/maps/axes.py [hunk 3]
```diff
@@ -1131,23 +1135,38 @@ def downsample(self, factor):
...
+                edges = self.edges[idx]
 
-            edges = self.edges[::factor]
-            return self.from_edges(edges, name=self.name, interp=self.interp)
+            else:
+                if np.mod(nbin, 1) > 0:
+                    raise ValueError(
+                        f"Number of {self.name} bins is not divisible by {factor}"
+                    )
+                edges = self.edges[::factor]
```
Comment (AtreyeeS) [inline gammapy/maps/axes.py:1149]: Do you still need this?
```
idx = np.arange(0, self.nbin, factor)
 if idx[-1] < self.nbin:
      idx = np.append(idx, self.nbin)
 edges = self.edges[idx]
```
should suffice.
The `ValueError` can be raised inside the `if` statement `if strict is True`

File: gammapy/maps/axes.py [hunk 4]
```diff
@@ -1131,23 +1135,38 @@ def downsample(self, factor):
...
+            downsampled = self.from_edges(edges, name=self.name, interp=self.interp)
+            return downsampled
         else:
             nbin = (self.nbin - 1) / factor
 
-            if np.mod(nbin, 1) > 0:
-                raise ValueError(
-                    f"Number of {self.name} bins - 1 is not divisible by {factor}"
-                )
+            if strict and np.mod(nbin, 1) > 0:
```
Comment (AtreyeeS) [inline gammapy/maps/axes.py:1155]: I don't think you need the separate handling for node type anymore either

File: gammapy/maps/tests/test_axes.py [hunk 5]
```diff
@@ -180,6 +180,26 @@ def test_up_downsample_consistency(factor):
...
+def test_strict_downsample():
+    axis1 = MapAxis.from_bounds(
+        10.0, 2e3, 6, interp="log", name="energy_true", unit="GeV"
+    )
+    with pytest.raises(ValueError) as exc_info:
+        axis1.downsample(4)
+    assert str(exc_info.value) == "Number of energy_true bins is not divisible by 4"
+
+    axis1_strict_down = axis1.downsample(4, strict=True)
+    assert_allclose(axis1_strict_down.nbin, 0.5 * (axis1.nbin) - 1)
```
Comment (AtreyeeS) [inline gammapy/maps/tests/test_axes.py:192]: What is this number you are checking against? It should be `(axis1.nbin)/factor + 1`, no?

**Review by AtreyeeS — CHANGES_REQUESTED** [review]
Thank you @Astro-Kirsty ! Left some comments inline. 
You can also adapt `get_edges_fixed_bins` in `estimators.utils`

File: gammapy/maps/axes.py (hunk 4 above)
Comment (AtreyeeS) [reply gammapy/maps/axes.py:1155]: On second thoughts, I think you do. To be able to preserve the `node_type`
But then, this logic needs modification

File: gammapy/maps/axes.py (hunk 3 above)
Comment (Astro-Kirsty) [reply gammapy/maps/axes.py:1148]: When running the tests, it appears I still need it

File: gammapy/maps/tests/test_axes.py (hunk 5 above)
Comment (Astro-Kirsty) [reply gammapy/maps/tests/test_axes.py:192]: That would be `(axis1.nbin)/4 + 1 = 2.5`

**Comment by AtreyeeS** [comment]
The current behaviour for `node_type=center` looks strange to me.
1. Why does it need `nbin-1` to be divisible by `factor` ? 
2. The rebinned axis is binned from the left, and the bins look strange - see last 2 lines of example 
3. Is it at all reasonable to downsample an axis defined at node centres?

```
from gammapy.maps import MapAxis
ax1 = MapAxis.from_nodes(np.linspace(0,3,6))

ax1.downsample(2) # Leads to an error - ValueError: Number of  bins - 1 is not divisible by 2

ax2 = ax1.downsample(5)
# ax2 now has 2 bins

# or
ax1 = MapAxis.from_nodes(np.linspace(0,9,10))
ax2= ax1.downsample(3)
# ax2 has 4 bins
ax2.edges # this starts from -1.5 whereas ax1.edges starts from -0.5

ax2.center # [0. 3. 6. 9.]
ax1.center # [0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]
```

Is there any motivation for this @registerrier @adonath ? Otherwise, I propose to forbid downsampling for node_type=center

**Comment by adonath** [comment]
> ax2.edges # this starts from -1.5 whereas ax1.edges starts from -0.5

Yes, this is because the bin size has increased, while the center has been kept the same. I hope this make sense.

**Comment by AtreyeeS** [comment]
Thanks @adonath for the clarification. I understand now that downsampling makes perfect sense, it's simply skipping every nth bin as you said. The `while preserving the axis limits` wording confused me.

I have kept the old behaviour.

**Review by registerrier — APPROVED** [review]
Thanks @AtreyeeS @Astro-Kirsty . This looks good to me.

**Pull request merged**